add non-catastrophic error MPIX_ERR_EAGAIN

raffenet opened the pull request.

raffenet commented:
test:jenkins/ch4/ofi

wesbland commented on src/include/mpi.h.in:
```diff
@@ -879,10 +879,11 @@ typedef int (MPIX_Grequest_wait_function)(int, void **, double, MPI_Status *);
 #define MPI_T_ERR_INVALID_NAME      73  /* Name doesn't match */
 #define MPI_T_ERR_INVALID           74  /* Generic error code for MPI_T added in MPI-3.1 */
 
+#define MPI_ERR_EAGAIN              75  /* Operation could not be issued */
```
Should this go below as `MPICH_ERR_EAGAIN` since it's not a constant in the MPI Standard?

wesbland commented on src/mpid/ch4/netmod/ofi/ofi_send.h:
```diff
@@ -30,7 +30,7 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_send_lightweight(const void *buf,
-        MPIDI_OFI_send_handler(MPIDI_Global.ctx[0].tx, buf, data_sz, NULL, comm->rank,
+        MPIDI_OFI_inject_handler(MPIDI_Global.ctx[0].tx, buf, data_sz, NULL, comm->rank,
                                MPIDI_OFI_av_to_phys(addr), match_bits,
                                NULL, MPIDI_OFI_DO_INJECT, MPIDI_OFI_CALL_LOCK);
```
We can probably drop the `MPIDI_OFI_DO_INJECT` if we're going to have separate calls for send/inject.

On the other hand, why does this need to be a separate function? Why can't it be in `MPIDI_OFI_send_handler`?

wesbland replied:
For that matter, why do we want to bubble `FI_EAGAIN` up to the user at all? If we're going to do that, I think we should do it after some limited number of retries since this is not an unexpected result from an OFI call. Can we just change the `MPIDI_OFI_CALL_RETRY` macro to do that?

raffenet replied:
Limited number of retries is probably a good idea. This is really just an experiment. Task-based runtimes that use MPI for communication might be able to context switch to something else if they get this type of error.

pavanbalaji replied:
Please use a CVAR for the number of retries, so users/tools can control it.

raffenet replied:
It's true that the send handler could also run into an eagain. The reason I left that as-is is because at that time, a send request has already been allocated. If we wanted to give up on that operation and return an error to the user, there's a little more cleanup to do first.

wesbland replied:
You could put it only in the inject part of that function.

raffenet commented:
I updated this to change the error class to `MPIX_`. This also needs to be modified to only return an error when the user has requested `MPI_ERRORS_RETURN`, meaning it will add a branch to the critical path. We probably need to add a configure switch for this to disable it by default if it is to be included in mainline.

wesbland commented on src/include/mpi.h.in:
```diff
@@ -879,10 +879,9 @@ typedef int (MPIX_Grequest_wait_function)(int, void **, double, MPI_Status *);
-
 #define MPI_ERR_LASTCODE    0x3fffffff  /* Last valid error code for a 
 					   predefined error class */
-#define MPICH_ERR_LAST_CLASS 74     /* It is also helpful to know the
```
This should go back to 74.

wesbland commented:
Why should we only bubble up the error during `MPI_ERRORS_RETURN`? We already have all of the error handling mechanisms to deal with this regardless of the current error handler.

raffenet commented:
Hmm, I suppose that's true. Still, we need to give users a way to opt-in to this error or else we'll start crashing programs that used to work.

pavanbalaji commented:
@raffenet wasn't the intent to only enable this when the user asks for it through a communicator info hint?

raffenet commented:
Yes, I believe that is in the proposal, but won't that add overhead even in the default case? For instance
```
if (comm->ncerr == false)
    call inject;
else
    call inject_with_eagain;
```
I don't want to drag down `ch4/ofi` performance to add this.

wesbland commented:
@pavanbalaji - Are you talking about a proposal from the FTWG? If so, there's no info key in that proposal.

@raffenet - I think you can just leave off the `if` entirely and let the normal error checking configuration take care of this.

raffenet commented:
@wesbland I'm not sure how? At some point we have to decide to give up and return to the user, as opposed to infinitely retrying like the current code.

pavanbalaji commented:
@wesbland: what Ken was referring to was the "proposed design", not the MPI Forum proposal.  I agree with @raffenet.  We cannot avoid that branch if we allow for both the infinite retry case and the limited retry case.

wesbland commented:
I see.

Even so, the branch should be only after getting `EAGAIN` right? Then it’s not a huge deal.

pavanbalaji commented on src/include/mpi.h.in:
```diff
@@ -879,7 +879,6 @@ typedef int (MPIX_Grequest_wait_function)(int, void **, double, MPI_Status *);
 #define MPI_T_ERR_INVALID_NAME      73  /* Name doesn't match */
 #define MPI_T_ERR_INVALID           74  /* Generic error code for MPI_T added in MPI-3.1 */
 
-
```
Random white space change.

pavanbalaji commented on src/mpid/ch4/netmod/ofi/ofi_impl.h:
```diff
@@ -115,6 +115,35 @@
+                                   FCNAME,                      \
+                                   fi_strerror(-_ret));      \
+    } while (0)
+
```
If you are using emacs, you can do `Ctrl-c-q` to reindent macros.

pavanbalaji replied:
Better yet, can this be made a `static inline` function?

raffenet commented:
@wesbland good point. In the optimal case we wouldn't touch any of the new code. Maybe it's not so bad to put this in the existing macro.

hajimefu commented on src/mpid/ch4/netmod/ofi/ofi_impl.h:
```diff
@@ -458,6 +487,31 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_send_handler(struct fid_ep *ep, const voi
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_inject_handler(struct fid_ep *ep, const void *buf,
+                                                      size_t len, void *desc, uint32_t dest,
```
I think `uint32_t dest` is the correct parameter name. (cf. #2990)

raffenet commented:
test:jenkins/ch4/ofi

raffenet commented:
test:jenkins/ch4/ofi

raffenet commented:
test:jenkins/ch4/ofi

raffenet commented:
test:jenkins/ch4/ofi

raffenet commented:
test:jenkins/ch4/ofi

raffenet commented:
test:jenkins/ch4/ofi

raffenet commented:
test:jenkins/ch4/ofi

raffenet commented:
@pavanbalaji @wesbland I made some modifications to the PR based on the comments.
1. fixed random whitespace issues
2. added communicator info hint to enable "eagain"-type errors
3. modified existing retry macro to handle both the limited and infinite retry cases. This could probably use a more sensible default and perhaps a CVAR to control the number of retries.

wesbland commented:
I don’t know if these test failures are ok or not. I’ll assume you’re dealing with those.

raffenet commented:
@wesbland good call. those failures shouldn't happen unless the user opt-ed in. I'll fix.

raffenet commented:
I had a very silly ternary operator bug before causing those failures.
test:jenkins/ch4/ofi

hajimefu commented on src/mpid/ch4/netmod/ofi/ofi_impl.h:
```diff
@@ -115,9 +115,10 @@
+#define MPIDI_OFI_CALL_RETRY(FUNC,STR,LOCK,EAGAIN)          \
     do {                                                    \
     ssize_t _ret;                                           \
+    int _retry = EAGAIN ? 1 : -1;                           \
```
Hmm, so this will add one branch in the critical path... how about making it configure-time option to enable/disable until `MPI_ERR_EAGAIN` becomes mandatory?

pavanbalaji replied:
Configure check seems like a really heavy hammer for this?  How about changing the caller to add the retry count instead?  We'd get that count from a CVAR, which would default to `-1`.

yfguo replied:
I have added the CVAR and moved the check out of the loop. So in the case where EAGAIN set to false, the retry loop won't pay the cost of one extra branch.

raffenet replied:
The conditional here is evaluating compile-time constant. It shouldn't actually be a penalty. That said, I agree that a CVAR is a good idea.

yfguo replied:
Not quite. Because for send, the EAGAIN is from the info hint of the communicator, which can only be evaluated at runtime. But I think the major concern was that we were checking EAGAIN in the following loop (which we no longer do in the latest patch).

raffenet replied:
Oh, right. I was thinking of a previous version. Sorry about that.

hajimefu replied:
:+1: Current version of the patch looks good. Thanks!

yfguo commented:
test:jenkins/ch4/ofi

yfguo commented on src/mpid/ch4/netmod/ofi/ofi_impl.h:
```diff
@@ -138,17 +138,16 @@
-    } while (_ret == -FI_EAGAIN);                           \
+        _retry--;                                           \
+    } while (_retry != -1 || !EAGAIN);                           \
+    MPIR_ERR_SETANDJUMP(mpi_errno,                          \
```
Yes, I am fixing it.

raffenet commented on src/mpid/ch4/netmod/ofi/ofi_impl.h:
```diff
@@ -138,17 +138,17 @@
-    } while (_ret == -FI_EAGAIN);                           \
+        _retry--;                                           \
+    } while (_retry != -1 || !EAGAIN);                           \
+    MPIR_ERR_CHKANDJUMP(_ret == -FI_EAGAIN && EAGAIN,       \
```
This is moving a branch that was previously protected by `if (likely(_ret==0)) break;`. It may now be more penalizing because of that.

yfguo replied:
Ah, that's true. I should move this `MPIR_ERR_CHKANDJUMP` back to its original place. So I guess the cost of it is avoidable in the retry loop.

yfguo commented:
test:jenkins/ch4/ofi

yfguo commented:
test:jenkins/ch4/ofi

yfguo commented:
Rebased

yfguo commented:
test:jenkins/ch4/ofi
